### ISSUE TITLE
Support needed for starred math equation environment

### ISSUE THREAD

**Issue by higham** [org2blog/org2blog#306]
At the moment, org2blog supports

\begin{equation}
...
\end{equation}

but not

\begin{equation*}
...
\end{equation*}

I'd love to use the latter, as when I export to LaTeX I mostly
don't want equations to be numbered (I need to export to both Wordpress and LaTeX).

I tried adding

               ((equal m "begin")
                (if (equal (match-string 2) "equation*")
                    (replace-match (concat "<p style=\"text-align:center\"> $latex "
                                           (substring (match-string 1) 16 -14)
                                           "$ </p>") nil t)))               

to ox-wp-latex-to-wp in ox-wp but it didn't work.
Probably the integer arguments 16 and -14 need tweaking.

**Comment by grettke** [comment]
For reference here is the current behavior.

Here is the Org2Blog file:

```
A

\begin{equation}
  1 + 1 = 2
\end{equation}

B

\begin{equation}
  3 * 3 = 9
\end{equation}

C

\begin{equation*}
54 * 7 = 378
\end{equation}

D

\begin{equation*}
18601646 - 57 = 18601589
\end{equation}
```

This is the LaTeX document it produces:

<img width="626" alt="Screen Shot 2021-04-19 at 3 09 17 PM" src="https://user-images.githubusercontent.com/2521647/115297127-70990000-a121-11eb-915a-bc7c4fbd7e70.png">

Note that the first environment _with_ the numbered equations is a `begin{equation}` and the second environment _without_ the numbered equations is a `begin{equation*}`. 

Here is the rendered HTML output on WordPress

<img width="514" alt="Screen Shot 2021-04-19 at 3 09 28 PM" src="https://user-images.githubusercontent.com/2521647/115297338-b5bd3200-a121-11eb-8476-8b6f30a64426.png">

**Comment by higham** [comment]
Thanks for looking into this, Grant.
As per LaTeX syntax, it should be

\begin{equation*}
54 * 7 = 378
\end{equation*}

with both occurrences of "equation" starred.

I got several compilation errors when I downloaded the latest from Melpa just now
 (my emacs-version = 27.1, org-version = 9.4.5):

In toplevel form:
wc-mode.el:75:1:Warning: defcustom for `wc-modeline-format' fails to specify
    type
wc-mode.el:75:1:Warning: defcustom for `wc-modeline-format' fails to specify
    type

In wc-count:
wc-mode.el:291:12:Warning: `interactive-p' is an obsolete function (as of
    23.2); use `called-interactively-p' instead.
wc-mode.el:290:26:Warning: `interactive-p' is an obsolete function (as of
    23.2); use `called-interactively-p' instead.

In toplevel form:
org2blog-test-system.el:43:7:Warning: assignment to free variable
    `org2blog/wp-blog-alist'

In end of data:
org2blog-test-system.el:384:1:Warning: the following functions are not known
    to be defined: org2blog-user-login, org2blog-buffer-post-save,
    org2blog-buffer-post-view, org2blog-buffer-post-publish,
    org2blog-buffer-post-trash, org2blog-buffer-page-save,
    org2blog-buffer-page-view, org2blog-buffer-page-publish,
    org2blog-buffer-page-trash, org2blog-subtree-post-save,
    org2blog-subtree-post-view, org2blog-subtree-post-publish,
    org2blog-subtree-post-trash, org2blog-subtree-page-save,
    org2blog-subtree-page-view, org2blog-subtree-page-publish,
    org2blog-subtree-page-trash

In toplevel form:
org2blog.el:265:1:Warning: Package cl is deprecated

In org2blog--create-categories:
org2blog.el:2321:28:Warning: `seq-contains' is an obsolete function (as of
    27.1); use `seq-contains-p' instead.

In org2blog--upload-files-replace-urls:
org2blog.el:2433:54:Warning: `org-plain-link-re' is an obsolete variable (as
    of Org 9.3); use `org-link-plain-re' instead.
org2blog.el:2734:1:Warning: Unused lexical variable `export-environment'

In org2blog--parse-subtree-entry:
org2blog.el:2785:52:Warning: `org-get-tags-at' is an obsolete function (as of
    Org 9.2); use `org-get-tags' instead.

**Comment by grettke** [comment]
About the equation environment: understood.

Did you update all of your packages along with Org2Blog, or only Org2Blog? The `wc-mode` that _hopefully_ Org2Blog isn't affecting that.

There are a lot of warnings. However, problem number one between my chair and keyboard is that I haven't updated Org2Blog to use the latest Emacs and Org mode. So I definitely can't reproduce your issue until I upgrade my local Emacs. 

To address what you are running into (as our others), I will update Org2Blog to the newer Emacs and Org mode. Once I've got that cut over, I'll report back and start back on this issue.

Have a good day.

**Comment by higham** [comment]
I updated all packages.
The message "Package cl is deprecated" is generated by lots of packages, not just Org2blog.

**Comment by grettke** [comment]
It looks like [`cl`](https://github.com/emacs-mirror/emacs/blob/master/etc/NEWS.27#L678) was deprecated and we should use `cl-lib` instead. A lot of packages are going to be updating eventually.

**Comment by grettke** [comment]
For the TLDR please jump to the end!

Really, these notes are pretty rough, but important.

Tonight I dug into this topic. The results are somewhat involved, but necessary to make sense of this, so I'm capturing them here. This all revolves around `equation` environment support in Org2Blog. In order to create a fix (somewhere) the entire workflow needs to get written down (for understanding, history, and potentially some kind of therapeutic effect)

How does WordPress handle `equation`? Short answer: it doesn't. As far as I can tell, WordPress's built in LaTeX support (that I'll call WPLaTeX) only handles one-line equations. Out of the box it has no clue. The next step to look is one higher: Org mode.

How does Org mode handle `equation`? Org mode handles [LaTeX fragments](https://orgmode.org/manual/LaTeX-fragments.html) using its built in library to export HTML. Exporting to plain old HTML is where to look for `equation` support. Using it is simple: Org recognizes *any* kind of environment or `$` math delimiters. Org mode can export LaTeX to HTML three ways. Find the details here:

The first way is: don't process fragments at all. Why? No clue but it is an option: all of your fragments are excluded from publishing.

The second way is to pass the fragments through verbatim. This makes a lot of sense because you delegate all of the processing work to something further down the chain. In the context of publishing to WordPress with HTML there are plenty of plugins for this.

The third way is for Org mode to an online converter for you named [MathJaX](https://www.mathjax.org/). This is the default. With this in mind consider how Org2Blog moves forward.

Org2Blog, in its exporter ~ox-wp~\`, has two ways to handle LaTeX HTML export:

-   Delegate work to Org mode that delegates the to MathJaX.
-   Perform the work itself in `ox-wp.el`. This is a custom exporter.

When delegating to Org mode:

Org mode's `ox-html` handles the exporting. `org-html-with-latex` defaults to `t` so it uses MathJaX. Out of the box MathJaX does not follow the `amsmath` equation behavior: it must be enabled. Without it the starred or unstarred `equation` environment doesn't use a number. For future reference [this documentation](https://docs.mathjax.org/en/latest/input/tex/eqnumbers.html) explains why and how. This is one solution to making numbering work correctly. The other is in `ox-wp` itself.

`ox-wp-latex-to-wp` does the custom exporting. It generates plain old HTML from scratch. The power here is that creates markup that WPLaTeX uses. That way when you install Org2Blog you get an entire LaTeX solution built in. The same problem that WPLaTeX reappears now though: WPLaTeX can't handle environments. Therefore `ox-wp` handles it and generates the code itself.

First it only handles a single line of the environment. That is not how it should work. Second, it doesn't include a number. Third, it doesn't account for numbers across equation environments. The solution is probably to implement something [like this](https://kitchingroup.cheme.cmu.edu/blog/2016/11/07/Better-equation-numbering-in-LaTeX-fragments-in-org-mode/).

The conclusion is the same as the TLDR:

-   WordPress's built in LaTeX can't handle the `equation` environment at all
-   Org2Blog's `ox-wp` has to add the ability to differentiate between starred and unstarred, keep track of the number of starred equations, and create the correct HTML
-   However (whoever) is using MathJaX needs to enable `ams` mode for its `equation` numbering. Once I figure out how to do this I'll update the docs and report back.

Consequently the easiest way to get correct numbering is to switch to MathJaX and once `ox-wp` adds support to switch back.

When that happens I'll comment in this ticket.

Thanks for your involvement and patience on this!

**Comment by higham** [comment]
Thank you very much for looking into this, Grant.
Since I first raised the request I've been using a macro which inserts

\begin{equation}\notag
\end{equation}

and I'm happily using that.
So I'm happy for things to stay as they are.

Ideally, I would use MathJax, but I don't have the plan that
allows plugins to be run.
